pkg/vmprovider/providers/vsphere2/vmlifecycle/update_status.go

```diff
@@ -457,3 +400,263 @@ func MarkBootstrapCondition(
+	// Otherwise, assign the network config to the VM's status.network.config
+	// field.
+	if reflect.DeepEqual(nc, emptyNetConfig) {
+		if vm.Status.Network != nil {
+			vm.Status.Network.Config = nil
+		}
+	} else {
+		if vm.Status.Network == nil {
+			vm.Status.Network = &vmopv1.VirtualMachineNetworkStatus{}
+		}
+		vm.Status.Network.Config = &nc
+	}
+}
+
+// updateGuestNetworkStatus updates the provided VM's status.network.config
+// field with information from the guestInfo.
+func updateGuestNetworkStatus(vm *vmopv1.VirtualMachine, gi *types.GuestInfo) {
+
+	if vm == nil {
+		panic("vm is nil")
+	}
+
+	if gi == nil {
+		return
+	}
+
+	var (
+		primaryIP4      string
+		primaryIP6      string
+		ifaceStatuses   []vmopv1.VirtualMachineNetworkInterfaceStatus
+		ipStackStatuses []vmopv1.VirtualMachineNetworkIPStackStatus
+	)
+
+	if ip := gi.IpAddress; ip != "" {
+		// Only act on the IP if it is valid.
+		if a := net.ParseIP(ip); len(a) > 0 {
+
+			// Ignore local IP addresses, i.e. addresses that are only valid on
+			// the guest OS. Please note this does not include private, or RFC
+			// 1918 (IPv4) and RFC 4193 (IPv6) addresses, ex. 192.168.0.2.
+			if !a.IsUnspecified() &&
+				!a.IsLinkLocalMulticast() &&
+				!a.IsLinkLocalUnicast() &&
+				!a.IsLoopback() {
+
+				if a.To4() != nil {
+					primaryIP4 = ip
+				} else {
+					primaryIP6 = ip
+				}
+			}
+		}
+	}
+
+	if len(gi.Net) > 0 {
+		var ifaceSpecs []vmopv1.VirtualMachineNetworkInterfaceSpec
+		if vm.Spec.Network != nil {
+			ifaceSpecs = vm.Spec.Network.Interfaces
+		}
+
+		slices.SortFunc(gi.Net, func(a, b types.GuestNicInfo) int {
+			// Sort by the DeviceKey (DeviceConfigId) to order the guest info
+			// list by the order in the initial ConfigSpec which is the order of
+			// the []ifaceSpecs since it is immutable.
+			return int(a.DeviceConfigId - b.DeviceConfigId)
+		})
+
+		ifaceIdx := 0
+		for i := range gi.Net {
+			deviceKey := gi.Net[i].DeviceConfigId
+
+			// Skip pseudo devices.
+			if deviceKey < 0 {
+				continue
+			}
+
+			var ifaceName string
+			if ifaceIdx < len(ifaceSpecs) {
+				ifaceName = ifaceSpecs[ifaceIdx].Name
+				ifaceIdx++
+			}
+
+			ifaceStatuses = append(
+				ifaceStatuses,
+				guestNicInfoToInterfaceStatus(
+					ifaceName,
+					deviceKey,
+					&gi.Net[i]))
+		}
+	}
+
+	if lip := len(gi.IpStack); lip > 0 {
+		ipStackStatuses = make([]vmopv1.VirtualMachineNetworkIPStackStatus, lip)
+		for i := range gi.IpStack {
+			ipStackStatuses[i] = guestIPStackInfoToIPStackStatus(&gi.IpStack[i])
+		}
+	}
+
+	var (
+		lip4 = len(primaryIP4)
+		lip6 = len(primaryIP6)
+		lip  = len(ipStackStatuses)
+		lis  = len(ifaceStatuses)
+	)
+
+	if lip4 > 0 || lip6 > 0 || lip > 0 || lis > 0 {
+		if vm.Status.Network == nil {
+			vm.Status.Network = &vmopv1.VirtualMachineNetworkStatus{}
+		}
+		if lip4 > 0 {
+			vm.Status.Network.PrimaryIP4 = primaryIP4
+		}
+		if lip6 > 0 {
+			vm.Status.Network.PrimaryIP4 = primaryIP4
```

Comment:
IP4 -> IP6 here (makes me think we're missing a test or assertion for this)

Reply:
We likely are. I can add one.